pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,345 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, LinearCombination, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_supported_obs = frozenset(["PauliZ", "Identity"])
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_measurements(tape, supported_base_obs=("PauliZ", "Identity")):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
+
+    Args:
+        tape (QNode or QuantumScript or Callable): The quantum circuit to modify the measurements of.
+        supported_base_obs (Optional, Iterable(Str)): A list of names of supported base observables.
+            Allowed names are 'PauliX', 'PauliY', 'PauliZ' and 'Hadamard'. If no list is provided,
```

Comment:
```suggestion
            Allowed names are ``'X'``, ``'Y'``, ``'Z'`` and ``'Hadamard'``. If no list is provided,
```
Note that `X`, `Y`, `Z` are the preferred entrypoints now

Reply:
Are we updating `qml.X(0).name` to be `"X"`? Or do you mean it would be preferable if `supported_base_obs` expected an iterable of classes instead of an iterable of observable names?

Reply:
> Are we updating qml.X(0).name to be "X"?

Not yet. Ok so you are thinking of this as exactly the `name` attribute of the observable, rather than just a general name for these strings? Do we expect the `name` attribute to be used when calling this transform?

Reply:
Right now, yes, it is, but can switch it to checking the class instead, its pretty much the same

Reply:
Updated to use classes instead of names, since the names might change :)